antenna/configlib.py

```
@@ -0,0 +1,323 @@
+
+    Namespace is prefixed, so key "foo" in namespace "bar" is ``FOO_BAR``.
+
+    """
+    def __init__(self, cfg):
+        self.cfg = cfg
+
+    def get(self, key, namespace=''):
+        if namespace:
+            key = '%s_%s' % (namespace, key)
+        key = key.upper()
+
+        if key in self.cfg:
+            return self.cfg[key]
+        return NO_VALUE
+
+
+class ConfigOSEnv(object):
+    """os.environ derived configuration layer
+
+    Namespace is prefixed, so key "foo" in namespace "bar" is ``FOO_BAR`` in
+    the ``os.environ``.
+
+    """
+    def get(self, key, namespace=''):
+        if namespace:
+            key = '%s_%s' % (namespace, key)
+
+        key = key.upper()
+        if key in os.environ:
+            return os.environ[key]
+
+        return NO_VALUE
+
+
+class ConfigIniEnv(object):
+    """.ini style configuration layer
+
+    Namespace is a config section. So "foo" in namespace "bar" is::
+
+        [bar]
+        foo=someval
+
+    """
+    def __init__(self, fn):
+        self._parser = ConfigParser()
+        self._parser.readfp(open(fn, 'r'))
+
+    def get(self, key, namespace='main'):
+        if self._parser.has_option(namespace, key):
+            return self._parser.get(namespace, key)
+        return NO_VALUE
+
+
+class ConfigManager(object):
+    """Manages multiple configuration environment layers"""
+    def __init__(self):
+        self.envs = self._initialize_environments()
+
+    def _initialize_environments(self):
+        # Add environments in order of precedence--first is the most important!
+        envs = []
+
+        # Add override environment
+        envs.append(ConfigOverrideEnv())
+
+        # Add OS environment
+        envs.append(ConfigOSEnv())
+
+        # Add .ini environment
+        fn = os.environ.get('ANTENNA_INI', 'antenna.ini')
+        if not os.sep in fn:
+            fn = os.path.join(os.getcwd(), fn)
+        if os.path.exists(fn):
+            envs.append(ConfigIniEnv(fn))
```

Here's a radical idea; change it from `fn = os.environ.get('ANTENNA_INI', 'antenna.ini')` to `fn = os.environ.get('ANTENNA_INI', NO_VALUE)`.
It circumgoes the problem of whether or not to raise or swallow the error. 

Yeah, logging is not going to be useful. But `warnings.warn()` is an option. It goes to stdout by default which is almost always read when the presence of the `antenna.ini` file is expected to exist. I.e. in local development.